### PR TITLE
Continuous Fuzzing Integration

### PR DESCRIPTION
**Pull request by yevgenypats**
Integrating a continuous fuzzing system (fuzzit) which will run long running jobs on releases and quick sanity tests via generated corpus every commit.

This is still work in progress and I changed the .travis.yml for testing purposes (I'll change it back to proper condition when the integration will be ready)

### CODE REVIEW

**Comment by ret2libc** [comment]
How is this different from https://github.com/google/oss-fuzz ?
How long will it take to run travis with this changes?

**Comment by yevgenypats** [comment]
Hi,  I still work on this so please don't merge yet.

1) it take about 2 min to build the targets so it doesn't affect the total build time
2) The oss fuzz integration is not currently working as discussed with @zlowram and pancake.
3) This adds an aditional value with not only running long running fuzzing job but also quick sanity tests on every commit that includes all the generated test-cases and crashes.

The integration should be ready very soon, already found a few bug. If this is not going to be merged please let me know as soon as possible as I invested quite some time for this integration because this is what we discussed in the group in telegram.

Cheers,

**Comment by ret2libc** [comment]
I'm absolutely ok with the idea. If @radare and @zlowram are ok as well I guess you are mostly good to go :)

I will wait until it's not WIP anymore to review.

**Comment by yevgenypats** [comment]
Thanks! I'll hope to finish this by tomorrow.

**Comment by Maijin** [comment]
but those fuzz tests are random or pre-defined?

**Comment by yevgenypats** [comment]
they are randomly generated using the long running fuzzing job but you have access to all the test-cases that were generated so you can reproduce the bugs if such occur during a sanity test run.

**Comment by Maijin** [comment]
But doesn't that mean that even for a patch that will only do a typo fix, the status of the Pull-request will be either green or red randomly?

**Comment by Maijin** [comment]
How can travis be useful to test for regressions with such a behavior?

**Comment by yevgenypats** [comment]
Hi,

The integration with Travis is only to start the job (So it won't block the CI). if there are bugs that occur this will send email to the relevant people + the commiter.

In anyway it won't introduce new bugs in a patch for a typo. If new test-case-bugs are generated using the long running jobs you will see in which commit it was introduced and which other commits are affected. This is pretty much the standard in OSSFuzz and in google Software afaik.

**Comment by Maijin** [comment]
Ok so if it's only starting a job on another website that's fine 👍 As long as it doesn't redify randomly that's all good to me

**Comment by yevgenypats** [comment]
Hi @ret2libc,

The Pull-Request is ready to merge though there is some more work in terms of bugfixes that the fuzzing already found and memory leaks (I'll be happy to help with those as well though I need someone with a better knowledge of radare2 code base).

Few things to note/do/to think about.

1) Can you please register at [(https://app.fuzzit.dev](https://app.fuzzit.dev) so I can add you to _radare2_ project.
2) Where should we send notifications to? do you have security@radare group or just send it to telegram using telegram bot?
3) The most important thing after merging I think will be working with someone on writing a better target (which is located at radare2-regressions) or understanding were the memory leaks are coming from and how to fix them because this prevent from a long running fuzzing to run successfully.

Cheers,
Yevgeny

**Comment by ret2libc** [comment]
> Can you please register at (https://app.fuzzit.dev so I can add you to radare2 project.
> Where should we send notifications to? do you have security@radare group or just send it to telegram using telegram bot?

@radare  ^

**Comment by ret2libc** [comment]
@yevgenypats where can we see right now the things found by fuzzit? how did you test?

**Comment by yevgenypats** [comment]
@ret2libc You need to access the dashboard to see the current stats. per the Segfault I have the file locally, I can open an Issue. I can make the dashboard publicly accessible if this is convenient though usually it should be only accessible to security teams but that really depends on the project.

**Comment by Maijin** [comment]
Personally I think that since we already have people reporting fuzz result on the issue tracker and some random people helping fix those. I wouldn't mind having them publicly available.

**Comment by yevgenypats** [comment]
@ret2libc @Maijin Ok, So I'll make those public and ping again after I test the permissions work as they should. probably tomorrow.

**Comment by Maijin** [comment]
That's just my opinion, maybe @radare wants them private I don't know

**Comment by yevgenypats** [comment]
@Maijin @ret2libc I  made the project publicly accessible (read only - you can't delete jobs, change settings etc...)

The fuzz results are located [here](https://app.fuzzit.dev/admin/2zv5qI33roZkRm0oO2Mi/target) I also linked to this url from README + added a badge that should indicate if fuzz sanity tests are succeeding/failing.

Few things to note

- I compiled the [ia_fuzz](https://github.com/radare/radare2-regressions/tree/master/fuzz/targets) without address sanitizer as there are [leaks](https://app.fuzzit.dev/admin/2zv5qI33roZkRm0oO2Mi/target/YVkkS6RPVpKhSixyFWcT/5egh3KYByw39sBALr0jH) which we need to open a new issue (I'll be happy to help but I need someone with a deeper knowledge of radare2 code base)
- The current status is that every commit/pull-request we will have the fuzz-target test all the [seed](https://github.com/radare/radare2-regressions/tree/master/fuzz/targets/corpora/ia_fuzz_seed_corpus) corpus and notify upon crashes.

I think this is a good start, after we merge we can start working on fixing the leaks so we will have a long running fuzzing jobs finding bugs for us:)

File: .travis.yml [hunk 1]
```diff
@@ -82,8 +93,18 @@ script:
   - export R2R_BRANCH=$([ -n "${TRAVIS_PULL_REQUEST_BRANCH}" ] && echo "${TRAVIS_PULL_REQUEST_BRANCH}" || echo "${R2R_DEFAULT_BRANCH}")
   - if [ "${R2R_BRANCH}" = master ]; then USER_NAME=radare; fi
   - export R2R_REPO=$(doc/repo PR_REGRESSIONS | sed "s/__USER__/${USER_NAME}/g")
-  - if [ "$TRAVIS_OS_NAME" != "osx" ]; then bash ./travis-extract-var.sh > ${TRAVIS_BUILD_DIR}/docker-env.sh ; cat ${TRAVIS_BUILD_DIR}/docker-env.sh ; fi
-  - if [ "$TRAVIS_OS_NAME" != "osx" ]; then docker run -v /home/travis/.ccache:/root/.ccache -v ${TRAVIS_BUILD_DIR}:${TRAVIS_BUILD_DIR} --entrypoint "/bin/bash" radareorg/r2-travis:latest -c ". /src/.nvm/nvm.sh; cd ${TRAVIS_BUILD_DIR}; . docker-env.sh; bash travis-script"; else bash travis-script; fi
+  - |
```
Comment (yevgenypats) [inline .travis.yml]: I just changed to multiline bash as YAML support this and I think this is more readable:)

**Comment by radare** [comment]
there's no such thing as security@radare you can just send those issues to my personal mail like its done for clusterfuzz but will be better if you can make that thing fill issues in github with the bins attached.

**Comment by yevgenypats** [comment]
@radare github issues are on the roadmap, hopefully soon.

**Comment by radare** [comment]
can you rebase to get it green and confirm this is ready to merge?

**Comment by yevgenypats** [comment]
@radare done. There is one job that is failing in Travis but I think this is failing also on master.

**Comment by radare** [comment]
Yes but thats a meson issue and the meson maintainers dont care at all about it. So you can consider it as green

**Comment by yevgenypats** [comment]
Got it. So you are merging?

**Comment by ret2libc** [comment]
Meson has nothing to do with master being red. It is red because people don’t fix asserts. It’s not related to this pr anyway.

File: .travis.yml [hunk 2]
```diff
@@ -82,8 +93,18 @@ script:
...
+    if [ "$TRAVIS_OS_NAME" != "osx" ]; then
+      bash ./travis-extract-var.sh > ${TRAVIS_BUILD_DIR}/docker-env.sh ; cat ${TRAVIS_BUILD_DIR}/docker-env.sh
```
Comment (radare) [inline .travis.yml]: Is this bash or posix shell? Can we use $SHELL instead and make this script portable?

File: .travis.yml [hunk 3]
```diff
@@ -82,8 +93,18 @@ script:
...
+    fi
+  - |
+    if [ "${FUZZIT}" == "1" ] ; then
+      docker run -v /home/travis/.ccache:/root/.ccache -v ${TRAVIS_BUILD_DIR}:${TRAVIS_BUILD_DIR} --entrypoint "/bin/bash" gcr.io/fuzzit-public/builders:llvm9.v1 -c "cd ${TRAVIS_BUILD_DIR}; . docker-env.sh; bash travis-script"
+    elif [ "$TRAVIS_OS_NAME" != "osx" ] ; then
+      docker run -v /home/travis/.ccache:/root/.ccache -v ${TRAVIS_BUILD_DIR}:${TRAVIS_BUILD_DIR} --entrypoint "/bin/bash" radareorg/r2-travis:latest -c ". /src/.nvm/nvm.sh; cd ${TRAVIS_BUILD_DIR}; . docker-env.sh; bash travis-script"
```
Comment (radare) [inline .travis.yml]: I would quote those vars to avoid problems in case the paths contain spaces

File: travis-script [hunk 4]
```diff
@@ -42,6 +42,27 @@ if [ "${INSTALL_SYSTEM}" == "meson" ] ; then
     ninja install || exit 1
     popd
     export PKG_CONFIG_PATH=$(pwd)/build/meson-private:${PKG_CONFIG_PATH}
+elif [ "${INSTALL_SYSTEM}" == "static" ]; then
+	echo "Compiling static with sys/static.sh"
+	./sys/static.sh
+	echo "Compiling Fuzzing Target"
+	export RADARE2_STATIC_BUILD=$(pwd)/r2-static
+	cd radare2-regressions/fuzz/targets
```
Comment (radare) [inline travis-script]: || exit 1

File: travis-script [hunk 5]
```diff
@@ -42,6 +42,27 @@ if [ "${INSTALL_SYSTEM}" == "meson" ] ; then
     ninja install || exit 1
     popd
     export PKG_CONFIG_PATH=$(pwd)/build/meson-private:${PKG_CONFIG_PATH}
+elif [ "${INSTALL_SYSTEM}" == "static" ]; then
+	echo "Compiling static with sys/static.sh"
+	./sys/static.sh
```
Comment (radare) [inline travis-script]: || exit 1

File: travis-script [hunk 6]
```diff
@@ -42,6 +42,27 @@ if [ "${INSTALL_SYSTEM}" == "meson" ] ; then
...
+	echo "Compiling Fuzzing Target"
+	export RADARE2_STATIC_BUILD=$(pwd)/r2-static
+	cd radare2-regressions/fuzz/targets
+	git remote -v
+	git branch
+	git rev-parse HEAD
+	# workaround for now
+	export LIB_FUZZING_ENGINE=" "
+	VERBOSE=1 make || exit 1
```
Comment (radare) [inline travis-script]: -j2 at least?

File: .travis.yml [hunk 7]
```diff
@@ -82,8 +93,18 @@ script:
...
+    else
+      bash travis-script
```
Comment (radare) [inline .travis.yml]: Juse ./travis-script and use proper hashbang and perms

File: .travis.yml (hunk 2 above)
Comment (yevgenypats) [reply .travis.yml]: I'm not sure regarding SHELL but I'll try. Anyway I didn't really change anything in this line apart from turning this into two lines (mult-line yaml string)

File: .travis.yml (hunk 7 above)
Comment (yevgenypats) [reply .travis.yml]: I didn't change this line as well but I'll try and remove bash to see if it doesn't brake anything. maybe it was intentional.

File: .travis.yml (hunk 2 above)
Comment (yevgenypats) [reply .travis.yml]: Done. Changed it to $SHELL

File: .travis.yml (hunk 3 above)
Comment (yevgenypats) [reply .travis.yml]: Done.

File: .travis.yml (hunk 7 above)
Comment (yevgenypats) [reply .travis.yml]: I changed to $SHELL as looks like it has to start in a new process or otherwise the build fails.

**Comment by yevgenypats** [comment]
@radare fixed the review. Thx.

**Comment by radare** [comment]
Thanks! Merging

**Comment by radare** [comment]
in the fuzzit website i can only get the type of error and the sample. is there a way to see the full asan crashlog like in clusterfuzz?

**Pull request merged**